rust/parquet/src/util/io.rs

```diff
@@ -31,47 +33,83 @@ pub trait Position {
 }
 
 /// Struct that represents a slice of a file data with independent start position and
-/// length. Internally clones provided file handle, wraps with BufReader and resets
-/// position before any read.
+/// length. Internally clones provided file handle, wraps with a custom implementation
+/// of BufReader that resets position before any read.
 ///
 /// This is workaround and alternative for `file.try_clone()` method. It clones `File`
 /// while preserving independent position, which is not available with `try_clone()`.
 ///
-/// Designed after `arrow::io::RandomAccessFile`.
+/// Designed after `arrow::io::RandomAccessFile` and `std::io::BufReader`
 pub struct FileSource<R: ParquetReader> {
-    reader: Mutex<BufReader<R>>,
-    start: u64, // start position in a file
-    end: u64,   // end position in a file
+    reader: RefCell<R>,
+    start: u64,     // start position in a file
+    end: u64,       // end position in a file
+    buf: Vec<u8>,   // the internal buffer `buf` in BufReader
+    buf_pos: usize, // equivalent to the `pos` param in BufReader
```

Comment:
This is not super useful as it requires ppl to jump into `BufReader` to check the documentation - maybe add additional comments on what these two fields are for?